test/extension/handlers/topologymutation/handler.go

```diff
@@ -0,0 +1,351 @@
+			infrav1.GroupVersion,
+			vmwarev1.GroupVersion,
+			controlplanev1.GroupVersion,
+			bootstrapv1.GroupVersion,
+		),
+	}
+}
+
+// GeneratePatches implements the HandlerFunc for the GeneratePatches hook.
+// The hook adds to the response the patches we are using in Cluster API E2E tests.
+// NOTE: custom RuntimeExtension must implement the body of this func according to the specific use case.
+func (h *ExtensionHandlers) GeneratePatches(ctx context.Context, req *runtimehooksv1.GeneratePatchesRequest, resp *runtimehooksv1.GeneratePatchesResponse) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("GeneratePatches is called")
+
+	// By using WalkTemplates it is possible to implement patches using typed API objects, which makes code
+	// easier to read and less error prone than using unstructured or working with raw json/yaml.
+	// IMPORTANT: by unit testing this func/nested func properly, it is possible to prevent unexpected rollouts when patches are modified.
+	topologymutation.WalkTemplates(ctx, h.decoder, req, resp, func(ctx context.Context, obj runtime.Object, variables map[string]apiextensionsv1.JSON, holderRef runtimehooksv1.HolderReference) error {
+		log := ctrl.LoggerFrom(ctx)
+
+		isControlPlane := holderRef.Kind == "KubeadmControlPlane"
+
+		switch obj := obj.(type) {
+		case *controlplanev1.KubeadmControlPlaneTemplate:
+			if err := patchKubeadmControlPlaneTemplate(ctx, obj, variables); err != nil {
+				log.Error(err, "Error patching KubeadmControlPlaneTemplate")
+				return errors.Wrap(err, "error patching KubeadmControlPlaneTemplate")
+			}
+		case *bootstrapv1.KubeadmConfigTemplate:
+			if err := patchKubeadmConfigTemplate(ctx, obj, variables); err != nil {
+				log.Error(err, "Error patching KubeadmConfigTemplate")
+				return errors.Wrap(err, "error patching KubeadmConfigTemplate")
+			}
+		case *infrav1.VSphereClusterTemplate:
+			if err := patchGovmomiClusterTemplate(ctx, obj, variables); err != nil {
+				log.Error(err, "Error patching VSphereClusterTemplate")
+				return errors.Wrap(err, "error patching VSphereClusterTemplate")
+			}
+		case *infrav1.VSphereMachineTemplate:
+			if err := patchGovmomiMachineTemplate(ctx, obj, variables, isControlPlane); err != nil {
+				log.Error(err, "Error patching VSphereMachineTemplate")
+				return errors.Wrap(err, "error patching VSphereMachineTemplate")
+			}
+		case *vmwarev1.VSphereClusterTemplate:
+			if err := patchSupervisorClusterTemplate(ctx, obj, variables); err != nil {
+				log.Error(err, "Error patching VSphereClusterTemplate")
+				return errors.Wrap(err, "error patching VSphereClusterTemplate")
+			}
+		case *vmwarev1.VSphereMachineTemplate:
+			if err := patchSupervisorMachineTemplate(ctx, obj, variables, isControlPlane); err != nil {
+				log.Error(err, "Error patching VSphereMachineTemplate")
+				return errors.Wrap(err, "error patching VSphereMachineTemplate")
+			}
+		}
+		return nil
+	}, topologymutation.PatchFormat{Format: runtimehooksv1.JSONMergePatchType})
```

Comment:
Let's add a comment why we need this

Reply:
Ah that's a good one :D

JSONPatch created a not working patch :-)

```
I0705 12:37:07.032800       1 engine.go:426] "Failed to apply patch with uid \"a9681ae7-37c4-4570-b4ba-596de6dd57e0\": error applying json patch (RFC6902): [{\"op\":\"replace\",\"path\":\"/spec/template/spec/controlPlaneEndpoint/host\",\"value\":\"192.168.26.227\"},{\"op\":\"replace\",\"path\":\"/spec/template/spec/controlPlaneEndpoint/port\",\"value\":6443},{\"op\":\"add\",\"path\":\"/spec/template/spec/identityRef\",\"value\":{\"kind\":\"Secret\",\"name\":\"k8s-upgrade-with-runtimesdk-mc7dnc\"}},{\"op\":\"add\",\"path\":\"/spec/template/spec/server\",\"value\":\"10.2.224.4\"},{\"op\":\"add\",\"path\":\"/spec/template/spec/thumbprint\",\"value\":\"D8:64:61:90:26:65:B8:1E:95:18:4B:68:56:D2:F7:34:FA:68:BF:2A\"}]: replace operation does not apply: doc is missing path: /spec/template/spec/controlPlaneEndpoint/host: missing value" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="k8s-upgrade-with-runtimesdk-8gq93l/k8s-upgrade-with-runtimesdk-mc7dnc" namespace="k8s-upgrade-with-runtimesdk-8gq93l" name="k8s-upgrade-with-runtimesdk-mc7dnc" reconcileID="1da34609-9aec-4b98-998a-e74429459848" patch="test-patch" uid="a9681ae7-37c4-4570-b4ba-596de6dd57e0"
```

The relevant part:
```json
  {
    "op": "replace",
    "path": "/spec/template/spec/controlPlaneEndpoint/host",
    "value": "192.168.26.227"
  },
  {
    "op": "replace",
    "path": "/spec/template/spec/controlPlaneEndpoint/port",
    "value": 6443
  },
```

Which created this error:
```
replace operation does not apply: doc is missing path: /spec/template/spec/controlPlaneEndpoint/host: missing value
```

Reply:
That is very bad @fabriziopandini 

Maybe we should switch to use JSONMergePatch per default. We should definitley let Sagar, etc. know

Reply:
will add it as comment.

Reply:
@fabriziopandini ^^ let's discuss this tomorrow in our sync